```cpumodel``` Crash report

### Description

Just trying the newest version.
```

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm is tainted, so must be replaced
-/+ resource "opennebula_virtual_machine" "vm" {
      ~ default_tags           = {} -> (known after apply)
      ~ gid                    = 0 -> (known after apply)
      ~ gname                  = "oneadmin" -> (known after apply)
      ~ id                     = "7" -> (known after apply)
      ~ ip                     = "x.x.x.x" -> (known after apply)
      ~ lcmstate               = 0 -> (known after apply)
        name                   = "vmk-dev-intel-01"
      ~ permissions            = "600" -> "660"
      ~ state                  = 6 -> (known after apply)
      ~ tags_all               = {} -> (known after apply)
      ~ template_disk          = [] -> (known after apply)
      ~ template_nic           = [] -> (known after apply)
      ~ template_section_names = {} -> (known after apply)
      ~ template_tags          = {} -> (known after apply)
      ~ uid                    = 0 -> (known after apply)
      ~ uname                  = "oneadmin" -> (known after apply)
        # (12 unchanged attributes hidden)

      ~ cpumodel {
          - model = "Skylake-Client-IBRS" -> null
        }

      ~ disk {
          + computed_cache           = (known after apply)
          ~ computed_dev_prefix      = "vd" -> (known after apply)
          + computed_discard         = (known after apply)
          ~ computed_driver          = "qcow2" -> (known after apply)
          + computed_io              = (known after apply)
          ~ computed_size            = 30720 -> (known after apply)
          ~ computed_target          = "vda" -> (known after apply)
          ~ computed_volatile_format = "qcow2" -> (known after apply)
          ~ disk_id                  = 0 -> (known after apply)
            # (3 unchanged attributes hidden)
        }
      ~ disk {
          + computed_cache           = (known after apply)
          ~ computed_dev_prefix      = "vd" -> (known after apply)
          + computed_discard         = (known after apply)
          ~ computed_driver          = "raw" -> (known after apply)
          + computed_io              = (known after apply)
          ~ computed_size            = 30720 -> (known after apply)
          ~ computed_target          = "vdb" -> (known after apply)
          ~ computed_volatile_format = "raw" -> (known after apply)
          ~ disk_id                  = 1 -> (known after apply)
            # (3 unchanged attributes hidden)
        }

      ~ graphics {
          ~ port          = "5907" -> (known after apply)
          - random_passwd = false -> null
            # (3 unchanged attributes hidden)
        }

      ~ nic {
          ~ computed_ip              = "x.x.x.x" -> (known after apply)
          ~ computed_mac             = "02:00:7a:5f:c3:90" -> (known after apply)
          ~ computed_model           = "virtio" -> (known after apply)
          + computed_physical_device = (known after apply)
          ~ computed_security_groups = [] -> (known after apply)
          ~ computed_virtio_queues   = "1" -> (known after apply)
          ~ network                  = "vlan500" -> (known after apply)
          - network_mode_auto        = false -> null
          ~ nic_id                   = 0 -> (known after apply)
          - security_groups          = [] -> null
            # (4 unchanged attributes hidden)
        }

      + template_section {
          + elements = {
              + "data" = "<memtune><hard_limit unit='MB'>256</hard_limit></memtune>"
              + "type" = "kvm"
            }
          + name     = "RAW"
        }

        # (2 unchanged blocks hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm: Destroying... [id=7]
module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm: Still destroying... [id=7, 10s elapsed]
module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm: Destruction complete after 15s
module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm: Creating...
╷
│ Error: Plugin did not respond
│ 
│   with module.vmk["vmk-dev-intel-01"].opennebula_virtual_machine.vm,
│   on ../single-vm/main.tf line 151, in resource "opennebula_virtual_machine" "vm":
│  151: resource "opennebula_virtual_machine" "vm" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-opennebula_v1.4.0 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 113 [running]:
github.com/OpenNebula/terraform-provider-opennebula/opennebula.generateVMTemplate(0xc00078e848?, 0xc0007201c8)
        github.com/OpenNebula/terraform-provider-opennebula/opennebula/shared_schemas.go:680 +0xd77
github.com/OpenNebula/terraform-provider-opennebula/opennebula.generateVm(0x0?, {0xc668e0?, 0xc0002f4940}, 0x0)
        github.com/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:2230 +0x35e
github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaVirtualMachineCreate({0xeb3848, 0xc0004c8180}, 0xc00032f500, {0xc668e0?, 0xc0002f4940})
        github.com/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_virtual_machine.go:350 +0x8b1
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc00045c620, {0xeb3880, 0xc00011f380}, 0xd?, {0xc668e0, 0xc0002f4940})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:707 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00045c620, {0xeb3880, 0xc00011f380}, 0xc0004111e0, 0xc00032ee00, {0xc668e0, 0xc0002f4940})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:837 +0xa7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00022b7a0, {0xeb3880?, 0xc00011f230?}, 0xc0002a00f0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/grpc_provider.go:1021 +0xe3c
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000339180, {0xeb3880?, 0xc00011e3f0?}, 0xc0007ba070)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xd2ce00?, 0xc000339180}, {0xeb3880, 0xc00011e3f0}, 0xc0007ba000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003661e0, {0xeb6b00, 0xc00050a1a0}, 0xc0007ac000, 0xc00046ab10, 0x13b6960, 0x0)
        google.golang.org/grpc@v1.56.3/server.go:1335 +0xde3
google.golang.org/grpc.(*Server).handleStream(0xc0003661e0, {0xeb6b00, 0xc00050a1a0}, 0xc0007ac000, 0x0)
        google.golang.org/grpc@v1.56.3/server.go:1712 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.56.3/server.go:947 +0xca
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.56.3/server.go:958 +0x15c

Error: The terraform-provider-opennebula_v1.4.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Terraform and Provider version

Terraform v1.5.5
on linux_amd64
+ provider registry.terraform.io/e-breuninger/netbox v3.5.1
+ provider registry.terraform.io/hashicorp/cloudinit v2.3.2
+ provider registry.terraform.io/hashicorp/dns v3.4.0
+ provider registry.terraform.io/hashicorp/vault v3.24.0
+ provider registry.terraform.io/ivoronin/macaddress v0.3.2
+ provider registry.terraform.io/opennebula/opennebula v1.4.0

### Affected resources and data sources

_No response_

### Terraform configuration

_No response_

### Expected behavior

Not crash...

### Actual behavior

Crashed

### Steps to Reproduce

set the cpumodel resource to ""

### Debug output

_No response_

### Panic output

_No response_

### Important factoids

_No response_

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 FEATURES:
 
 * resources/opennebula_virtual_network: allow to modify the user owning the resource (#529)
+* resources/opennebula_virtual_machine: add nil checks before type casting (#530)
 
 # 1.4.0 (January 22nd, 2024)
 

--- a/opennebula/shared_schemas.go
+++ b/opennebula/shared_schemas.go
@@ -602,9 +602,11 @@ func makeNICVector(nicConfig map[string]interface{}) *shared.NIC {
 func addOS(tpl *vm.Template, os []interface{}) {
 
 	for i := 0; i < len(os); i++ {
-		osconfig := os[i].(map[string]interface{})
-		tpl.AddOS(vmk.Arch, osconfig["arch"].(string))
-		tpl.AddOS(vmk.Boot, osconfig["boot"].(string))
+		if os[i] != nil {
+			osconfig := os[i].(map[string]interface{})
+			tpl.AddOS(vmk.Arch, osconfig["arch"].(string))
+			tpl.AddOS(vmk.Boot, osconfig["boot"].(string))
+		}
 	}
 
 }
@@ -612,30 +614,32 @@ func addOS(tpl *vm.Template, os []interface{}) {
 func addGraphic(tpl *vm.Template, graphics []interface{}) {
 
 	for i := 0; i < len(graphics); i++ {
-		graphicsconfig := graphics[i].(map[string]interface{})
+		if graphics[i] != nil {
+			graphicsconfig := graphics[i].(map[string]interface{})
 
-		for k, v := range graphicsconfig {
+			for k, v := range graphicsconfig {
 
-			if isEmptyValue(reflect.ValueOf(v)) {
-				continue
+				if isEmptyValue(reflect.ValueOf(v)) {
+					continue
+				}
+
+				switch k {
+				case "listen":
+					tpl.AddIOGraphic(vmk.Listen, v.(string))
+				case "type":
+					tpl.AddIOGraphic(vmk.GraphicType, v.(string))
+				case "port":
+					tpl.AddIOGraphic(vmk.Port, v.(string))
+				case "keymap":
+					tpl.AddIOGraphic(vmk.Keymap, v.(string))
+				case "passwd":
+					tpl.AddIOGraphic(vmk.Passwd, v.(string))
+				case "random_passwd":
+					// Convert bool to string
+					tpl.AddIOGraphic(vmk.RandomPassword, map[bool]string{true: "YES", false: "NO"}[v.(bool)])
+				}
+
 			}
-
-			switch k {
-			case "listen":
-				tpl.AddIOGraphic(vmk.Listen, v.(string))
-			case "type":
-				tpl.AddIOGraphic(vmk.GraphicType, v.(string))
-			case "port":
-				tpl.AddIOGraphic(vmk.Port, v.(string))
-			case "keymap":
-				tpl.AddIOGraphic(vmk.Keymap, v.(string))
-			case "passwd":
-				tpl.AddIOGraphic(vmk.Passwd, v.(string))
-			case "random_passwd":
-				// Convert bool to string
-				tpl.AddIOGraphic(vmk.RandomPassword, map[bool]string{true: "YES", false: "NO"}[v.(bool)])
-			}
-
 		}
 
 	}
@@ -648,6 +652,10 @@ func generateVMTemplate(d *schema.ResourceData, tpl *vm.Template) error {
 	log.Printf("Number of disks: %d", len(disks))
 
 	for i := 0; i < len(disks); i++ {
+		if disks[i] == nil {
+			continue
+		}
+
 		diskconfig := disks[i].(map[string]interface{})
 
 		// ConflictsWith can't be used among attributes of a nested part: disk, nic etc.
@@ -677,17 +685,21 @@ func generateVMTemplate(d *schema.ResourceData, tpl *vm.Template) error {
 	//Generate CPU Model definition
 	cpumodel := d.Get("cpumodel").([]interface{})
 	for i := 0; i < len(cpumodel); i++ {
-		cpumodelconfig := cpumodel[i].(map[string]interface{})
-		tpl.CPUModel(cpumodelconfig["model"].(string))
+		if cpumodel[i] != nil {
+			cpumodelconfig := cpumodel[i].(map[string]interface{})
+			tpl.CPUModel(cpumodelconfig["model"].(string))
+		}
 	}
 
 	//Generate VM Group definition
 	vmgroup := d.Get("vmgroup").([]interface{})
 	for i := 0; i < len(vmgroup); i++ {
-		vmgconfig := vmgroup[i].(map[string]interface{})
-		vmgroupTpl := tpl.AddVector("VMGROUP")
-		vmgroupTpl.AddPair("VMGROUP_ID", vmgconfig["vmgroup_id"].(int))
-		vmgroupTpl.AddPair("ROLE", vmgconfig["role"].(string))
+		if vmgroup[i] != nil {
+			vmgconfig := vmgroup[i].(map[string]interface{})
+			vmgroupTpl := tpl.AddVector("VMGROUP")
+			vmgroupTpl.AddPair("VMGROUP_ID", vmgconfig["vmgroup_id"].(int))
+			vmgroupTpl.AddPair("ROLE", vmgconfig["role"].(string))
+		}
 	}
 
 	vmcpu, ok := d.GetOk("cpu")
@@ -728,10 +740,12 @@ func generateVMTemplate(d *schema.ResourceData, tpl *vm.Template) error {
 	//Generate RAW definition
 	raw := d.Get("raw").([]interface{})
 	for i := 0; i < len(raw); i++ {
-		rawConfig := raw[i].(map[string]interface{})
-		rawVec := tpl.AddVector("RAW")
-		rawVec.AddPair("TYPE", rawConfig["type"].(string))
-		rawVec.AddPair("DATA", rawConfig["data"].(string))
+		if raw[i] != nil {
+			rawConfig := raw[i].(map[string]interface{})
+			rawVec := tpl.AddVector("RAW")
+			rawVec.AddPair("TYPE", rawConfig["type"].(string))
+			rawVec.AddPair("DATA", rawConfig["data"].(string))
+		}
 	}
 
 	descr, ok := d.GetOk("description")

